Source/VMware.PSDesiredStateConfiguration/Classes/DscItems.ps1

```diff
@@ -254,7 +254,11 @@ class DscConfigurationCompiler {
     DscConfigurationCompiler([string] $ConfigName, [Hashtable] $CustomParams, [Hashtable] $ConfigurationData) {
         $this.ConfigName = $ConfigName
         $this.CustomParams = $CustomParams
-        $this.ConfigurationData = $ConfigurationData
+
+        # configurationData gets cloned because it's state gets mutated during execution.
+        if ($null -ne $ConfigurationData) {
+            $this.ConfigurationData = $ConfigurationData.Clone()
```

Comment:
The **Clone()** does a shallow copy as described [here](https://docs.microsoft.com/en-us/dotnet/api/system.collections.hashtable.clone?view=net-5.0). I think a deep copy should be created.